File: sources/zig/11_parallel_sum/main.zig
// task 11 parallel_sum — expected output: 7500000075000000
// build: zig build-exe -O ReleaseFast main.zig    run: ./main
// deviation: Zig 0.16 removed std.posix.write, so the line goes out through std.Io.File.stdout().

const std = @import("std");

const range: u64 = 25_000_000;

const Worker = struct {
    t: u64,
    result: u64 = 0,

    fn run(self: *Worker) void {
        var acc: u64 = 0;
        var i: u64 = self.t * range;
        const end: u64 = (self.t + 1) * range;
        while (i < end) : (i += 1) {
            switch (i % 4) {
                0 => acc += 1,
                1 => acc += i,
                2 => acc += 2 * i,
                3 => acc += 3 * i,
                else => unreachable,
            }
        }
        self.result = acc;
    }
};

pub fn main(init: std.process.Init) !void {
    const io = init.io;

    var workers: [4]Worker = undefined;
    var threads: [4]std.Thread = undefined;
    for (0..4) |t| {
        workers[t] = .{ .t = @intCast(t) };
        threads[t] = try std.Thread.spawn(.{}, Worker.run, .{&workers[t]});
    }
    for (threads) |thread| thread.join();

    var total: u64 = 0;
    for (workers) |worker| total += worker.result;

    var line_buf: [32]u8 = undefined;
    const line = try std.fmt.bufPrint(&line_buf, "{d}\n", .{total});
    try std.Io.File.stdout().writeStreamingAll(io, line);
}

File: sources/pascal/11_parallel_sum/main.pas
{ task 11 parallel_sum — expected output: 7500000075000000 }
{ build: fpc -O3 -oprogram main.pas    run: ./program }
{ Windows x64: i386-win32 fpc + cross.x86_64-win64 add-on, build with -Px86_64; run as program.exe }
{ note: cthreads is pulled in on Unix so the RTL installs the POSIX thread manager. }

program prog;
{$mode objfpc}{$H+}

uses
  {$IFDEF UNIX}cthreads,{$ENDIF}
  SysUtils, Classes;

type
  TWorker = class(TThread)
  private
    FLo, FHi: Int64;
    FPartial: Int64;
  protected
    procedure Execute; override;
  public
    constructor Create(ALo, AHi: Int64);
    property Partial: Int64 read FPartial;
  end;

constructor TWorker.Create(ALo, AHi: Int64);
begin
  FLo := ALo;
  FHi := AHi;
  FPartial := 0;
  inherited Create(True);   { suspended, so the range is in place before Execute runs }
end;

procedure TWorker.Execute;
var
  i, acc: Int64;
begin
  acc := 0;
  for i := FLo to FHi - 1 do
  begin
    case i mod 4 of
      0: Inc(acc);
      1: Inc(acc, i);
      2: Inc(acc, 2 * i);
      3: Inc(acc, 3 * i);
    end;
  end;
  FPartial := acc;
end;

var
  workers: array[0..3] of TWorker;
  i: Integer;
  total: Int64;
begin
  for i := 0 to 3 do
    workers[i] := TWorker.Create(Int64(i) * 25000000, (Int64(i) + 1) * 25000000);
  for i := 0 to 3 do
    workers[i].Start;      { all four run at the same time }
  total := 0;
  for i := 0 to 3 do
  begin
    workers[i].WaitFor;
    total := total + workers[i].Partial;
    workers[i].Free;
  end;
  WriteLn(total);
end.
